File: bench/clbg_c_ref/fannkuch.c
// CLBG fannkuch-redux — C reference (sequential).
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

static void solve(int n, long *out_max, long *out_check) {
    long perm[32], perm1[32], count[32];
    for (int i = 0; i < n; i++) perm1[i] = i;
    for (int i = 0; i < n; i++) count[i] = 0;
    long max_flips = 0, checksum = 0, permutation_count = 0;
    int r = n;
    for (;;) {
        while (r != 1) { count[r-1] = r; r--; }
        for (int i = 0; i < n; i++) perm[i] = perm1[i];
        long flips = 0;
        long k = perm[0];
        while (k != 0) {
            int j = 0; int m = (int)k;
            while (j < m) {
                long tmp = perm[j]; perm[j] = perm[m]; perm[m] = tmp;
                j++; m--;
            }
            flips++;
            k = perm[0];
        }
        if (flips > max_flips) max_flips = flips;
        if ((permutation_count & 1) == 0) checksum += flips;
        else checksum -= flips;
        permutation_count++;

        for (;;) {
            if (r == n) { *out_max = max_flips; *out_check = checksum; return; }
            long p0 = perm1[0];
            for (int i = 0; i < r; i++) perm1[i] = perm1[i+1];
            perm1[r] = p0;
            count[r]--;
            if (count[r] > 0) break;
            r++;
        }
    }
}

int main(int argc, char **argv) {
    int n = (argc > 1) ? atoi(argv[1]) : 12;
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    long m, c;
    solve(n, &m, &c);
    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed_ms = (t1.tv_sec - t0.tv_sec) * 1000.0 + (t1.tv_nsec - t0.tv_nsec) / 1.0e6;
    printf("fannkuch N=%d max=%ld checksum=%ld\n", n, m, c);
    printf("elapsed: %.0fms\n", elapsed_ms);
    return 0;
}

File: bench/clbg_rust_ref/src/fannkuch.rs
// CLBG fannkuch-redux — Rust reference (sequential, no parallelism).
// Same algorithm as the Soma version: enumerate permutations,
// count flips, track max + checksum.

use std::env;
use std::time::Instant;

fn solve(n: usize) -> (i64, i64) {
    let mut perm = vec![0i64; n];
    let mut perm1: Vec<i64> = (0..n as i64).collect();
    let mut count = vec![0i64; n];
    let mut max_flips = 0i64;
    let mut checksum = 0i64;
    let mut r = n;
    let mut permutation_count = 0i64;

    loop {
        while r != 1 {
            count[r - 1] = r as i64;
            r -= 1;
        }
        for i in 0..n {
            perm[i] = perm1[i];
        }
        let mut flips = 0i64;
        let mut k = perm[0];
        while k != 0 {
            let mut j = 0;
            let mut m = k as usize;
            while j < m {
                perm.swap(j, m);
                j += 1;
                m -= 1;
            }
            flips += 1;
            k = perm[0];
        }
        if flips > max_flips { max_flips = flips; }
        if permutation_count % 2 == 0 {
            checksum += flips;
        } else {
            checksum -= flips;
        }
        permutation_count += 1;

        loop {
            if r == n {
                return (max_flips, checksum);
            }
            let p0 = perm1[0];
            for i in 0..r {
                perm1[i] = perm1[i + 1];
            }
            perm1[r] = p0;
            count[r] -= 1;
            if count[r] > 0 { break; }
            r += 1;
        }
    }
}

fn main() {
    let n: usize = env::args().nth(1).map(|s| s.parse().unwrap()).unwrap_or(12);
    let t0 = Instant::now();
    let (m, c) = solve(n);
    let elapsed = t0.elapsed();
    println!("fannkuch N={} max={} checksum={}", n, m, c);
    println!("elapsed: {:?}", elapsed);
}
